File: BM24_二叉树的中序遍历/java_solutions/SolutionTest.java
package BM24_二叉树的中序遍历.java_solutions;

import java.util.*;

public class SolutionTest {

    public static void main(String[] args) {
        // 创建所有的实现类
        List<Solution> solutions = Arrays.asList(
                new RecursiveSolution(),
                new StackSolution(),
                new BetterStackSolution());

        // 对每一个Solution进行测试
        for (Solution solution : solutions) {
            System.out.println("Testing with solution: " + solution.getClass().getSimpleName());

            // Test Case 1: Empty tree
            TreeNode root1 = null;
            int[] result1 = solution.inorderTraversal(root1);
            int[] expected1 = {};
            System.out.println("Test Case 1: " + (Arrays.equals(result1, expected1) ? "passed" : "failed"));

            // Test Case 2: Single node tree
            TreeNode root2 = new TreeNode(1);
            int[] result2 = solution.inorderTraversal(root2);
            int[] expected2 = { 1 };
            System.out.println("Test Case 2: " + (Arrays.equals(result2, expected2) ? "passed" : "failed"));

            // Test Case 3: Left-skewed tree
            TreeNode root3 = new TreeNode(3);
            root3.left = new TreeNode(2);
            root3.left.left = new TreeNode(1);
            int[] result3 = solution.inorderTraversal(root3);
            int[] expected3 = { 1, 2, 3 };
            System.out.println("Test Case 3: " + (Arrays.equals(result3, expected3) ? "passed" : "failed"));

            // Test Case 4: Right-skewed tree
            TreeNode root4 = new TreeNode(1);
            root4.right = new TreeNode(2);
            root4.right.right = new TreeNode(3);
            int[] result4 = solution.inorderTraversal(root4);
            int[] expected4 = { 1, 2, 3 };
            System.out.println("Test Case 4: " + (Arrays.equals(result4, expected4) ? "passed" : "failed"));

            // Test Case 5: Full binary tree
            TreeNode root5 = new TreeNode(4);
            root5.left = new TreeNode(2);
            root5.right = new TreeNode(6);
            root5.left.left = new TreeNode(1);
            root5.left.right = new TreeNode(3);
            root5.right.left = new TreeNode(5);
            root5.right.right = new TreeNode(7);
            int[] result5 = solution.inorderTraversal(root5);
            int[] expected5 = { 1, 2, 3, 4, 5, 6, 7 };
            System.out.println("Test Case 5: " + (Arrays.equals(result5, expected5) ? "passed" : "failed"));

            // Test Case 6: Mixed tree
            TreeNode root6 = new TreeNode(5);
            root6.left = new TreeNode(3);
            root6.right = new TreeNode(8);
            root6.left.left = new TreeNode(1);
            root6.left.right = new TreeNode(4);
            root6.right.left = new TreeNode(7);
            int[] result6 = solution.inorderTraversal(root6);
            int[] expected6 = { 1, 3, 4, 5, 7, 8 };
            System.out.println("Test Case 6: " + (Arrays.equals(result6, expected6) ? "passed" : "failed"));

            System.out.println(); // 为了分隔每种实现的结果
        }
    }
}

File: BM24_二叉树的中序遍历/python_solutions/solution_test.py
from solution import TreeNode, Solution
# Helper function to build a tree from a list (level-order)
def build_tree(values):
    if not values:
        return None
    nodes = [TreeNode(x) if x is not None else None for x in values]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids:
                node.left = kids.pop()
            if kids:
                node.right = kids.pop()
    return root

# Test cases
def test_inorder_traversal():
    solution = Solution()
    
    # Test Case 1: Empty Tree
    root1 = None
    result1 = solution.inorderTraversal(root1)
    expected1 = []
    print("Test Case 1 - Empty Tree:", "Passed" if result1 == expected1 else f"Failed (Result: {result1})")

    # Test Case 2: Single Node
    root2 = TreeNode(1)
    result2 = solution.inorderTraversal(root2)
    expected2 = [1]
    print("Test Case 2 - Single Node:", "Passed" if result2 == expected2 else f"Failed (Result: {result2})")

    # Test Case 3: Complete Binary Tree
    # Tree structure:
    #       1
    #      / \
    #     2   3
    #    / \
    #   4   5
    root3 = build_tree([1, 2, 3, 4, 5])
    result3 = solution.inorderTraversal(root3)
    expected3 = [4, 2, 5, 1, 3]
    print("Test Case 3 - Complete Binary Tree:", "Passed" if result3 == expected3 else f"Failed (Result: {result3})")

    # Test Case 4: Left-Skewed Tree
    # Tree structure:
    #     1
    #    /
    #   2
    #  /
    # 3
    root4 = build_tree([1, 2, None, 3])
    result4 = solution.inorderTraversal(root4)
    expected4 = [3, 2, 1]
    print("Test Case 4 - Left-Skewed Tree:", "Passed" if result4 == expected4 else f"Failed (Result: {result4})")

    # Test Case 5: Right-Skewed Tree
    # Tree structure:
    # 1
    #  \
    #   2
    #    \
    #     3
    root5 = build_tree([1, None, 2, None, 3])
    result5 = solution.inorderTraversal(root5)
    expected5 = [1, 2, 3]
    print("Test Case 5 - Right-Skewed Tree:", "Passed" if result5 == expected5 else f"Failed (Result: {result5})")

    # Test Case 6: Unbalanced Tree
    # Tree structure:
    #       1
    #      / 
    #     2   
    #      \
    #       3
    root6 = build_tree([1, 2, None, None, 3])
    result6 = solution.inorderTraversal(root6)
    expected6 = [2, 3, 1]
    print("Test Case 6 - Unbalanced Tree:", "Passed" if result6 == expected6 else f"Failed (Result: {result6})")

# Run tests
test_inorder_traversal()

